### PR TITLE
TYP: overload lib.maybe_convert_objects

### PR DESCRIPTION
**Pull request by jbrockmendel**

### CODE REVIEW

File: pandas/_libs/lib.pyi [hunk 1]
```diff
@@ -51,23 +53,65 @@ def is_float_array(values: np.ndarray, skipna: bool = False): ...
...
+
+@overload
+def maybe_convert_objects(
+    objects: np.ndarray,  # np.ndarray[object]
+    try_float: bool = ...,
+    safe: bool = ...,
+    convert_datetime: Literal[False] = False,
+    convert_timedelta: bool = ...,
+    convert_to_nullable_integer: Literal[True] = ...,
 ) -> ArrayLike: ...
```
Comment (MarcoGorelli) [inline pandas/_libs/lib.pyi:85]: Nice...the reason the overload works here (with both `Literal[True]` and `Literal[False]` taking defaults) is that the return type of the second one is a supertype of the return type of the first one. Haven't tried running this yet, but it looks correct to me - if you don't specify `convert_to_nullable_integer`, it should match the first overload

Comment (jbrockmendel) [reply pandas/_libs/lib.pyi:85]: im not clear on what you're asking for here

Comment (MarcoGorelli) [reply pandas/_libs/lib.pyi:85]: nothing :) I was just commenting in case it wasn't clear why here you were able to let both overload variants take defaults but in most other cases we could only let one of the two take defaults

File: pandas/core/strings/object_array.py [hunk 2]
```diff
@@ -61,9 +61,12 @@ def _str_map(self, f, na_value=None, dtype: Optional[Dtype] = None):
...
         if not len(self):
-            # error: Argument 1 to "ndarray" has incompatible type "int";
-            # expected "Sequence[int]"
-            return np.ndarray(0, dtype=dtype)  # type: ignore[arg-type]
+            # error: Argument "dtype" to "array" has incompatible type
+            # "Union[ExtensionDtype, str, dtype[Any], Type[object]]"; expected
+            # "Union[dtype[Any], None, type, _SupportsDType, str,
+            # Union[Tuple[Any, int], Tuple[Any, Union[int, Sequence[int]]], List[Any],
+            # _DTypeDict, Tuple[Any, Any]]]"
+            return np.array([], dtype=dtype)  # type: ignore[arg-type]
```
Comment (MarcoGorelli) [inline pandas/core/strings/object_array.py:69]: I'm confused about why `mypy` reports this, if I try making a file `t.py` with
```python
import numpy as np
from pandas._typing import Dtype, Optional

def foo(dtype: Optional[Dtype] = None):
    if dtype is None:
        dtype = np.dtype("object")
    return (np.array([], dtype=type))
```
then `mypy` type checks it just fine. Do you know what I might be missing?

**Review by MarcoGorelli — APPROVED** [review]
There's a part (which I've commented on) where I haven't understood why `mypy` is reporting an error - on than that, this looks good to me

File: pandas/core/strings/object_array.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/strings/object_array.py:69]: no idea.  i think this edit is unrelated to most of the rest; using `np.ndarray` here instead of `np.array`  weirds me out

Comment (simonjayhawkins) [reply pandas/core/strings/object_array.py:69]: The `Dtype` alias includes `ExtensionDtype` which cannot be passed on to numpy

```
>>> np.array([], dtype=pd.Int64Dtype())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: Cannot interpret 'Int64Dtype()' as a data type
>>> 
```

> Do you know what I might be missing?

`return (np.array([], dtype=type))` -> `return (np.array([], dtype=dtype))`

Comment (MarcoGorelli) [reply pandas/core/strings/object_array.py:69]: I'm embarrassed, thanks!

File: pandas/core/arrays/string_arrow.py [hunk 3]
```diff
@@ -420,10 +420,8 @@ def fillna(self, value=None, method=None, limit=None):
         if mask.any():
             if method is not None:
                 func = missing.get_fill_func(method)
-                # error: Argument 1 to "to_numpy" of "ArrowStringArray" has incompatible
-                # type "Type[object]"; expected "Union[str, dtype[Any], None]"
                 new_values, _ = func(
-                    self.to_numpy(object),  # type: ignore[arg-type]
+                    self.to_numpy("object"),
```
Comment (simonjayhawkins) [inline pandas/core/arrays/string_arrow.py:424]: This is an unrelated change to the scope (from the PR title) of this PR. The mypy error is a false positive, no need to change code. will be fixed in #41185

but I guess nbd (other than merge conflicts)

File: pandas/core/strings/object_array.py (hunk 2 above)
Comment (simonjayhawkins) [reply pandas/core/strings/object_array.py:69]: so this looks like there maybe an actual bug here. As part of the ArrowStringArray work, I have been parameterising existing tests on object dtype arrays with StringArray and ArrowStringArray and this has maybe uncovered some latent bugs with StringArray.

I think OK to leave this ignore as a 'fix later' and out of scope here.

File: pandas/core/tools/datetimes.py [hunk 4]
```diff
@@ -393,18 +389,14 @@ def _convert_listlike_datetimes(
             require_iso8601 = not infer_datetime_format
             format = None
 
-    # error: Incompatible types in assignment (expression has type "None", variable has
-    # type "ExtensionArray")
-    result = None  # type: ignore[assignment]
+    result = None
```
Comment (simonjayhawkins) [inline pandas/core/tools/datetimes.py:392]: is this needed? The `if result is None` will always be True?

**Review by simonjayhawkins — COMMENTED** [review]
Thanks @jbrockmendel generally lgtm

**Review by simonjayhawkins — APPROVED** [review]
Thanks @jbrockmendel lgtm
